### ISSUE TITLE
hooks-callbacks

### ISSUE THREAD

**Issue by JavierClearImageAI** [facebookresearch/detectron2#867]
## How to implement some callbacks?

I need to implement ***lrt_finder***, ***early_stopping***, and ***model_checkpoint*** ("best" accuracy if possible. Otherwise, minimum loss).

1. Is it possible to get these callbacks?:

NOTE:

1. I have checked fastai, but since all the model is built in torch, I would prefer to use pure detectron or just modify the minimum possible. I guess we could get trainer.model, trainer.loss, etc and introduce this data on the lrt_finder or other callbacks?

2. Also, any hint on how to build these callbacks using [hooks?](https://github.com/facebookresearch/detectron2/blob/master/detectron2/engine/hooks.py). I just have a couple of days to implement all three to detectron training, so I would like to check out the most efficient-fast way.

This seems like a nice starting point: https://detectron2.readthedocs.io/tutorials/extend.html

I hope I made myself clear. Thanks a lot

**Comment by JavierClearImageAI** [comment]
ok, for the moment this is how I'm going to quick-solve it:

the output folder contains a "metrics.json" file which contains all the information I need, like the accuracy, the learning rates, the losses every 20 iterations, etc. I will clean it each training (since it seems to accumulate multiple trainings there), then I will train a small subset multiple times and analyze the results, in order to get at least, the optimal number of iterations and the right lrt for my dataset. A more interesting way to do this would be to use the hooks to build our custom callbacks.

**Comment by JavierClearImageAI** [comment]
just adding:  

    cfg.SOLVER.CHECKPOINT_PERIOD = 20

to the default training, we get all the solvers every 20 iterations.
Then we can select the best one based on the "metrics.json" which is in the outputs directory.

I just found out that we could create our own default trainer based on the [detectron2.engine.defaults.py]((https://github.com/facebookresearch/detectron2/blob/7fb05ed0f644214f1d631540ae7b6dffdb3c92e2/detectron2/engine/defaults.py)),
similarly like how they do in  [detectron2/tools/train_net.py ](https://github.com/facebookresearch/detectron2/blob/5e2a6f62ef752c8b8c700d2e58405e4bede3ddbe/tools/train_net.py)

**Comment by ppwwyyxx** [comment]
It's always easier to use your own loop to implement custom training logic, one example is https://github.com/facebookresearch/detectron2/blob/master/tools/plain_train_net.py.

You can abstract these logic into hooks but it's often more work. For example, for early stopping and saving the best model, you can obtain the metrics from `self.trainer.storage` (no need to parse the json file). 
I don't know how `lrt_finder` works but I suppose you'll need to run training a few times - so it's not about hooks.
